pandas/_libs/tslibs/offsets.pyx

```diff
@@ -456,8 +456,12 @@ cdef class BaseOffset:
         return type(self)(n=1, normalize=self.normalize, **self.kwds)
 
     def __add__(self, other):
-        if util.is_array(other) and other.dtype == object:
-            return np.array([self + x for x in other])
+        if util.is_array(other):
+            from pandas.api.types import is_datetime64_any_dtype
+            if other.dtype == object:
+                return np.array([self + x for x in other])
+            elif is_datetime64_any_dtype(other):
+                return self._apply_array(other).astype(other.dtype)
```

Comment:
Do we need the astype here?

Reply:
I found that without this DateOffset._apply_array did not respect the dtype was passed in
once added in, I could use e.g. datetime64[D] and get a datetime64[D] back

I'll do some further testing when working on unit tests

Reply:
FWIW `datetime64[D]` is not a supported unit in pandas